src/jemalloc.c

```diff
@@ -3310,13 +3317,26 @@ ixallocx_prof(tsd_t *tsd, void *ptr, size_t old_usize, size_t size,
 		 */
 		thread_event(tsd, usize - usize_max);
 	}
+
+	/*
+	 * At this point we can still safely get the original profiling
+	 * information associated with the ptr, because (a) the extent_t object
+	 * associated with the ptr still lives and (b) the profiling info
+	 * fields are not touched.  "(a)" is asserted in the outer je_xallocx()
+	 * function, and "(b)" is indirectly verified below by checking that
+	 * the alloc_tctx field is unchanged.
+	 */
+	prof_info_t prof_info;
 	if (usize == old_usize) {
+		prof_info_get(tsd, ptr, alloc_ctx, &prof_info);
 		prof_alloc_rollback(tsd, tctx, false);
-		return usize;
+	} else {
+		prof_info_get_and_reset(tsd, ptr, alloc_ctx, &prof_info);
```

Comment:
move this into `prof_realloc`, like what we have for `prof_free`. Then the `prof_info_get()` can be done before the `if ` above.

Reply:
This might not work for `rallocx()`. The reason is as follows:

In, `xallocx()`, the original allocation is always kept, at least in terms of it's corresponding `extent_t object, so `prof_realloc()` can safely access it (and thus obtain `prof_info` and reset the extent). This is similar to `free()`, where the original allocation is kept while `prof_free()` is executed. So, for `xallocx()` / `free()`, the extent reset logic can be safely put into `prof_realloc()` / `prof_free()`, and strictly speaking the `prof_info` design is not needed at all.

However, `rallocx()` is different: at the time `prof_realloc()` is entered, the original extent may or may not be there, so it's not possible to do either `prof_info` fetch or extent reset in `prof_realloc()`. The original allocation might have been gone in the core `iralloct()` call.

One way to resolve it is to ingest the `prof_info` fetch and extent reset somewhere below the `iralloct()` layer, but I feel it's complicated, both in terms of code change and in terms of code structure (since the core jemalloc logic has to have additional dependencies).

The other way, which is what I'm following now, is to do the `prof_info` fetch and extent reset early for `rallocx()`. In such a way the code change is simpler and the profiling layer can still be kept light and on top of the core jemalloc layer. The only seemingly wrong part is that, if `rallocx()` ends up not changing the allocation at all, I would still count it as being freed and then allocated. However, this actually aligns with our current mindset on `rallocx()`: a no-move case has always been treated as a free-then-allocate, from both the `thread.(de)allocated` perspective and the profiling perspective. See that `prof_realloc()` calls `prof_malloc_sample_object()` and `prof_free_sampled_object()` without checking `moved`, and that `rallocx()` never does `prof_alloc_rollback()` unless OOM. These are all different from `xallocx()`. Also see discussions on #1691.